web/src/server/api/routers/comments.ts

```diff
@@ -0,0 +1,238 @@
+import { z } from "zod";
+
+import { throwIfNoProjectAccess } from "@/src/features/rbac/utils/checkProjectAccess";
+import {
+  createTRPCRouter,
+  protectedProjectProcedure,
+} from "@/src/server/api/trpc";
+import { CommentObjectType } from "../../../../../packages/shared/dist/prisma/generated/types";
+import { Prisma, CreateCommentData } from "@langfuse/shared";
+import { auditLog } from "@/src/features/audit-logs/auditLog";
+import { TRPCError } from "@trpc/server";
+
+const COMMENT_OBJECT_TYPE_TO_PRISMA_MODEL = {
+  [CommentObjectType.TRACE]: "trace",
+  [CommentObjectType.OBSERVATION]: "observation",
+  [CommentObjectType.SESSION]: "session",
+  [CommentObjectType.PROMPT]: "prompt",
+} as const;
+
+const validateCommentReferenceObject = async ({
+  ctx,
+  input,
+}: {
+  ctx: any;
+  input: z.infer<typeof CreateCommentData>;
+}): Promise<void> => {
+  const { objectId, objectType, projectId } = input;
+  const prismaModel = COMMENT_OBJECT_TYPE_TO_PRISMA_MODEL[objectType];
+
+  if (!prismaModel) {
+    throw new Error(`No prisma model for object type ${objectType}`);
+  }
+
+  const model = ctx.prisma[prismaModel];
+  const object = await model.findFirst({
+    where: {
+      id: objectId,
+      projectId,
+    },
+  });
+
+  if (!object) {
+    throw new Error(
+      `No ${prismaModel} with id ${objectId} in project ${projectId}`,
+    );
+  }
+};
```

Comment:
Would love if this fct was more typesafe, but unsure if that's achievable. Open to feedback :)

Reply:
mh, this is creative. I'd probably just skip it and not validate the reference object, but now that it exists I'd keep it as long as Max does not have a strong opinion here

Reply:
Max's happy with the function 👍